solidity/random-beacon/README.adoc

```diff
@@ -1 +1,324 @@
+pool and receive a fixed, governable reward for reporting DKG timeout. The
+reward is paid from the DKG reward pool. 
+
+[[dkg-submit-eligibility]]
+The order in which operators are supposed to submit a DKG result is enforced
+on-chain. The first member eligible to submit the DKG result is a member with
+index `keccak256(new_group_pubkey) % group_size`. Then after a governable
+submission eligibility delays, members with subsequent indices are becoming
+eligible one after another.
+
+[NOTE]
+For example, if `hash(new_group_pubkey) % group_size = 62`, `group_size = 64`,
+group members are becoming eligible in the following order:
+`62, 63, 64, 1, 2, 3, 4, 5, 6, 7, 8, 9, ..., 61`. 
+
+Governance needs to ensure that both the frequency of relay requests with which
+a new DKG starts and the fee for a single request make DKG result submitter reward
+attractive enough so that no one wants to bail out from submitting the result.
+
+The transaction submitting relay entry is not reimbursable and implementation
+ensures the gas cost of this transaction is as low as possible, below 200k gas
+when no callback is executed.
+
+Everyone is eligible to submit relay entry at any time but off-chain clients are
+expected to agree and follow the following order to minimize the gas cost and
+distribute costs: the first group member eligible to submit the result is
+`new_entry % group_size`; then, if the selected member does not provide an entry
+within the governable eligibility period, `(new_entry % group_size) + 1` and
+so on.
+
+If some group members are notoriously ignoring their duty, the group can vote on
+failed <<heartbeats,heartbeat>> notification for these operators.
+
+T rewards will be distributed continuously to all operators registered in the beacon
+sortition pool, excluding operators who were marked as ineligible for rewards
+due to failing the heartbeat.
+
+[[heartbeats]]
+=== Heartbeats
+
+Off-chain clients are free to execute any heartbeat protocol they want to ensure
+group members are alive and nodes are operating properly.
+
+[TIP]
+One example of a heartbeat protocol is signing some piece of information every
+nth blocks and first making sure the information cannot be used for
+`RandomBeacon.reportUnauthorizedSigning()`, that is, the signed information can
+not become `msg.sender` for `reportUnauthorizedSigning` call.
+
+Group members can agree upon members that failed the heartbeat and issue a
+heartbeat failure claim. If the required threshold of group members signed
+the heartbeat failure claim, they can submit it to
+`RandomBeacon.notifyFailedHeartbeat(Heartbeat.FailureClaim calldata claim, uint256 nonce)`
+function and have the group members who failed the heartbeat excluded from
+the sortition pool rewards for a governable time period.
+
+The submitter of the failed heartbeat claim receives a reward from a separate
+notifier reward pool, funded by DAO  for heartbeat failure claims specifically.
```

Comment:
double space between "DAO" and "for"

Reply:
https://github.com/keep-network/keep-core/pull/2746/commits/e156cf8d6ad16d3cdb3879d940d9a59a6eee16db